File: pset2/crack.c
#define _XOPEN_SOURCE

#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <unistd.h>

int main(int argc, string argv[])
{
    if(argc != 2)
    {
        printf("Usage: %s k\n", argv[0]);
        return 1;
    }

    char salt[2];
    strncpy(salt, argv[1], 2);

    char letters[53];
    letters[0] = '\0';
    int index = 1;
    for (char i = 'A'; i <= 'Z'; i++)
    {
        letters[index++] = i;
    }
    for (char i = 'a'; i <= 'z'; i++)
    {
        letters[index++] = i;
    }

    char suspect[5];
    for(int i = 0; i < 53; i++)
    {
        for (int j = 0; j < 53; j++)
        {
            for (int k = 0; k < 53; k++)
            {
                for (int l = 1; l < 53; l++)
                {
                    suspect[0] = letters[l];
                    suspect[1] = letters[k];
                    suspect[2] = letters[j];
                    suspect[3] = letters[i];

                    string hash = crypt(suspect, salt);
                    if (strcmp(hash, argv[1]) == 0)
                    {
                        printf("%s\n", suspect);
                        return 0;
                    }
                }
            }
        }
    }
    printf("Nothing found...\n");
}

File: pset7/crack/crack.py
import sys
import crypt


def char_range(c1, c2):
    yield from range(ord(c1), ord(c2) + 1)


if len(sys.argv) != 2:
    print(f"Usage: {sys.argv[0]}\n")
    sys.exit(1)


salt = sys.argv[1]
letters = [None] * 53
letters[0] = '.'
index = 1

for i in char_range('A', 'Z'):
    letters[index] = chr(i)
    index += 1

for i in char_range('a', 'z'):
    letters[index] = chr(i)
    index += 1

for i in range(53):
    for j in range(53):
        for k in range(53):
            for l in range(53):
                password = letters[l] + letters[k] + letters[j] + letters[i]
                print(f"Password: {password}")
                hash = crypt.crypt(password, salt)
                if (hash == sys.argv[1]):
                    print(suspect)
                    sys.exit(0)

print("Nothing found...\n")